src/EditorFeatures/Core/Implementation/CallHierarchy/Finders/AbstractCallFinder.cs

```diff
@@ -48,51 +45,58 @@ public void CancelSearch()
             _cancellationSource.Cancel();
         }
 
-        public void StartSearch(CallHierarchySearchScope searchScope, ICallHierarchySearchCallback callback)
+        public void StartSearch(SymbolKey symbolKey, Project project, CallHierarchySearchScope searchScope, ICallHierarchySearchCallback callback)
         {
-            Task.Run(() => SearchAsync(callback, searchScope, _cancellationSource.Token), _cancellationSource.Token);
+            var asyncToken = _asyncListener.BeginAsyncOperation(this.GetType().Name + ".Search");
+
+            Task.Run(() => SearchAsync(symbolKey, project, searchScope, callback, _cancellationSource.Token), _cancellationSource.Token)
+                .CompletesAsyncOperation(asyncToken);
         }
 
-        private async Task SearchAsync(ICallHierarchySearchCallback callback, CallHierarchySearchScope scope, CancellationToken cancellationToken)
+        private async Task SearchAsync(SymbolKey symbolKey, Project project, CallHierarchySearchScope scope, ICallHierarchySearchCallback callback, CancellationToken cancellationToken)
         {
-            callback.ReportProgress(0, 1);
+            // The error message to show if we had an error. null will mean we succeeded.
+            string completionErrorMessage = null;
 
-            var asyncToken = _asyncListener.BeginAsyncOperation(this.GetType().Name + ".Search");
+            try
+            {
+                callback.ReportProgress(0, 1);
 
-            // Follow the search task with task that lets the callback know we're done and which
-            // marks the async operation as complete.  Note that we pass CancellationToken.None
-            // here.  That's intentional.  This operation is *not* cancellable.
+                var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+                var resolution = symbolKey.Resolve(compilation, cancellationToken: cancellationToken);
 
-            var workspace = _project.Solution.Workspace;
-            var currentProject = workspace.CurrentSolution.GetProject(_project.Id);
-            var compilation = await currentProject.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
-            var resolution = _symbol.Resolve(compilation, cancellationToken: cancellationToken);
+                var documents = this.Documents ?? IncludeDocuments(scope, project);
 
-            var documents = this.Documents ?? IncludeDocuments(scope, currentProject);
+                var currentSymbol = resolution.Symbol;
 
-            var currentSymbol = resolution.Symbol;
+                if (currentSymbol == null)
+                {
+                    throw new Exception(string.Format(WorkspacesResources.The_symbol_0_cannot_be_located_within_the_current_solution, SymbolName));
```

Comment:
:bulb: Cleaner to just set `completionErrorMessage` and return

Reply:
If there was a way I could make completionErrorMessage only visible to the catch/finally blocks I'd do so. :frowning:

Reply:
💭 You already hoisted the variable out of the try block, so it should be accessible for setting and returning (which would then get picked up in the `finally` block). Is there something else I'm missing?

Reply:
I guess I'm considering that variable to be an implementation detail that the try shouldn't touch.

Reply:
An acceptable alternative would be calling `callback.SearchFailed` directly, and then returning. You can do this by moving the `SearchSucceeded` call up into the try block after the last `ReportProgress` call.

Reply:
I just rewrote this how I thought I should have done it originally. :smile: